orderer/consensus/smartbft/chain.go

```diff
@@ -455,6 +467,118 @@ func (c *BFTChain) HandleMessage(sender uint64, m *smartbftprotos.Message, metad
+		if err := proto.Unmarshal(aux, committeeFeedback); err != nil {
+			c.Logger.Warningf("Failed unmarshaling committeeFeedback, error: %v", err)
+			continue
+		}
+		list := committeeFeedback.Suspects
+		cleanList := removeDuplicates(list)
+		allSuspects = append(allSuspects, cleanList...)
+	}
+	return agreedSuspects(allSuspects, 1) // TODO use f
+}
+
+func (c *BFTChain) getSuspectsFromBlock(proposal *smartbftprotos.Proposal) (bool, []int32) {
+	tuple := &ByteBufferTuple{}
+	if err := tuple.FromBytes(proposal.Payload); err != nil {
+		c.Logger.Warningf("Failed reading proposal payload, error: %v", err)
+		return false, nil
+	}
+	metadata := &common.BlockMetadata{}
+	if err := proto.Unmarshal(tuple.B, metadata); err != nil {
+		c.Logger.Warningf("Failed unmarshaling block metadata, error: %v", err)
+		return false, nil
+	}
+	if metadata == nil || len(metadata.Metadata) < len(common.BlockMetadataIndex_name) {
+		c.Logger.Warningf("Block metadata is either missing or contains too few entries")
+		return false, nil
+	}
+	signatureMetadata := &common.Metadata{}
+	if err := proto.Unmarshal(metadata.Metadata[common.BlockMetadataIndex_SIGNATURES], signatureMetadata); err != nil {
+		c.Logger.Warningf("Failed unmarshaling block signature metadata, error: %v", err)
+		return false, nil
+
+	}
+	ordererMDFromBlock := &common.OrdererBlockMetadata{}
+	if err := proto.Unmarshal(signatureMetadata.Value, ordererMDFromBlock); err != nil {
+		c.Logger.Warningf("Failed unmarshaling orderer block metadata, error: %v", err)
+		return false, nil
+	}
+	return true, ordererMDFromBlock.HeartbeatSuspects
+}
+
+func (c *BFTChain) verifySuspects(prp *smartbftprotos.PrePrepare) bool {
+	agreedSuspects := c.getSuspectsFromSignatures(prp.PrevCommitSignatures)
+	success, blockSuspects := c.getSuspectsFromBlock(prp.Proposal)
+	if !success {
+		c.Logger.Warningf("Couldn't read suspects from block")
+		return false
+	}
+
+	rtc := c.RuntimeConfig.Load().(RuntimeConfig)
+	if rtc.isConfig {
+		if len(blockSuspects) != 0 {
+			c.Logger.Warningf("Last block was a config block, but the suggested suspects list is not empty")
+			return false
+		}
+		c.Logger.Infof("Last block was a config block, therefore disregarding the suspects verification")
+		return true
+	}
+
+	if len(agreedSuspects) != len(blockSuspects) {
+		c.Logger.Warningf("Length of suspects list don't match, according to the signatures the length should be %d, while in the block the length is %d", len(agreedSuspects), len(blockSuspects))
+		return false
+	}
+	for i, s := range agreedSuspects {
+		if blockSuspects[i] != s {
+			c.Logger.Warningf("Suspect doesn't match, according to the signatures the %d'th suspect should be %d, while in the block the suspect is %d", i, s, blockSuspects[i])
+			return false
+		}
+	}
+	return true
+}
+
+func setupHeartbeatMonitor(selfID uint64, currentCommittee []committee.Node, allNodes []uint64) (myRole Role, heartbeatSenders []uint64, heartbeatReceivers []uint64) {
```

Comment:
this looks to me like a function that is easy to test in a UT, it also doesn't really setup the HBM but more like determine its configuration.